Qualcomm AI Engine Direct - Upload QC tutorial. (#6915)
google-ai-edge/LiteRT: litert/vendors/qualcomm/README.md

      - ${QAIRT}/lib/x86_64-linux-clang/libQnnHtp.so
      - ${QAIRT}/lib/x86_64-linux-clang/libQnnPrepare.so
      - ${QAIRT}/lib/x86_64-linux-clang/libQnnSystem.so

### Specify other QAIRT version in LiteRT
Modify ${LiteRT}/third_party/qairt/workspace.bzl and change `strip_prefix`. And setup environment variable `LITERT_QAIRT_SDK`.

For example, the original workspace.bzl is using "qairt/2.42.0.251225" and users want to use 2.44, then the workspace.bzl should be changed from:
```bash
def qairt():
    configurable_repo(
        name = "qairt",
        build_file = "@//third_party/qairt:qairt.BUILD",
        local_path_env = "LITERT_QAIRT_SDK",
        strip_prefix = "qairt/2.42.0.251225",
        url = "https://softwarecenter.qualcomm.com/api/download/software/sdks/Qualcomm_AI_Runtime_Community/All/2.42.0.251225/v2.42.0.251225.zip",
        file_extension = "zip",
    )
```
To:
```bash
def qairt():
    configurable_repo(
        name = "qairt",
        build_file = "@//third_party/qairt:qairt.BUILD",
        local_path_env = "LITERT_QAIRT_SDK",
        strip_prefix = "qairt/2.44.0.260225",

>         url = "https://softwarecenter.qualcomm.com/api/download/software/sdks/Qualcomm_AI_Runtime_Community/All/2.42.0.251225/v2.42.0.251225.zip",

Reviewer: ```suggestion
        url = "https://softwarecenter.qualcomm.com/api/download/software/sdks/Qualcomm_AI_Runtime_Community/All/2.44.0.260225/v2.44.0.260225.zip",
```

Author: Thanks for the reminder! Also update the above description.

        file_extension = "zip",
    )
```

## Bazel Flow (Compile & Execute)

Build "apply_plugin_main" compilation tool and compile the model on Linux workstation (Ubuntu-22.04):

### Steps
```bash
# Build apply_plugin_main with bazel

> export LITERT_QAIRT_SDK=<PATH_TO_LITERT_REPO>/third_party/qairt/

Reviewer: If we export `LITERT_QAIRT_SDK`, do we also need to modify `workspace.bzl`?

Author: Thanks for the reminder! There's no need to export this when using released SDK, let me remove it.

# Qualcomm Vendor Integration

> ## Enviroment Setup

Reviewer 2: ```suggestion
## Environment Setup
```

> You can follow <PATH_TO_LITERT_REPO>/g3doc/instructions/BUILD_INSTRUCTIONS.md to setup environment, the instructions below are what we used in our developement environment.

Reviewer 2: ```suggestion
You can follow <PATH_TO_LITERT_REPO>/g3doc/instructions/BUILD_INSTRUCTIONS.md to setup environment, the instructions below are what we used in our development environment.
```

> 0. Operation System

Reviewer 2: ```suggestion
0. Operating System
```

   - Ubuntu 22.04

1. Bazel
   - Download .deb file from [Bazelisk release](https://github.com/bazelbuild/bazelisk/releases) and install.
```bash
# for x86_64 Linux
wget https://github.com/bazelbuild/bazelisk/releases/download/v1.28.1/bazelisk-amd64.deb

> sudo dkpg -i ./bazelisk-amd64.deb

Reviewer 2: ```suggestion
sudo dpkg -i ./bazelisk-amd64.deb
```

Please specify the home path of the Android SDK to use. [Default is /usr/<UserName>/Android/Sdk]: /path/to/android-sdk-30/

Please specify the Android SDK API level to use. [Available levels: ['30']] [Default is 30]: 30

Please specify an Android build tools version to use. [Available versions: ['30.0.3']] [Default is 30.0.3]: 30.0.3

Preconfigured Bazel build configs. You can use any of the below by adding "--config=<>" to your build command. See .bazelrc for more details.
        --config=mkl            # Build with MKL support.
        --config=mkl_aarch64    # Build with oneDNN and Compute Library for the Arm Architecture (ACL).
        --config=monolithic     # Config for mostly static monolithic build.
        --config=numa           # Build with NUMA support.
        --config=dynamic_kernels        # (Experimental) Build kernels into separate shared objects.
        --config=v1             # Build with TensorFlow 1 API instead of TF 2 API.
Preconfigured Bazel build configs to DISABLE default on features:
        --config=nogcp          # Disable GCP support.
        --config=nonccl         # Disable NVIDIA NCCL support.
```

6. CMake

>    - Currently we use 4.0.1, please set up this for "CMake Flow (Compile & Execute)" section.

Reviewer 3: Need modification if section title "CMake Flow (Compile & Execute)" changed.

```bash
wget https://github.com/Kitware/CMake/releases/download/v4.0.1/cmake-4.0.1-linux-x86_64.tar.gz
tar -xzvf cmake-4.0.1-linux-x86_64.tar.gz
```

## QAIRT SDK

### Download link
- https://softwarecenter.qualcomm.com/api/download/software/sdks/Qualcomm_AI_Runtime_Community/All/2.44.0.260225/v2.44.0.260225.zip

### Official Document
- https://docs.qualcomm.com/doc/80-63442-10/topic/general_overview.html

### Interface Header Files
- ${QAIRT}/include/QNN/

### Required Libraries

> Here're some standard concepts of "How QNN works on the Qulacomm devices" users should know:

Reviewer 2: ```suggestion
Here're some standard concepts of "How QNN works on the Qualcomm devices" users should know:
```

  - Backend
       - Currently, LiteRT QC compiler plugin supports the following QNN backends:
      - Htp
      - Dsp
      - Ir

>       - Saver

Reviewer 3: ```suggestion
      - Currently, LiteRT QC compiler plugin supports the following QNN backends:
      - - Htp
      - - Dsp
      - - Ir
      - - Saver
```

  - Platform

>     - Here're the supported platforms of Qualcomm devices:

Reviewer 3: ```suggestion
    - Here're the supported operating system (OS) platforms of Qualcomm devices:
```

      - x86_64-linux-clang > Linux work station
      - aarch64-android > Android device

>       - Ubuntu-24.04 & aarch64-oe-linux-gcc11.2 &  > IoT device

Reviewer 3: ```suggestion
    - Here're the supported platforms of Qualcomm devices:
      - aarch64-android: for Android devices.
      - x86_64-linux-clang: for x86 Linux devices.
      - x86_64-windows-msvc: for x86 Windows devices.
      - aarch64-oe-linux-gcc11.2: for aarch64 Linux devices (IoT devices).
      - aarch64-windows-msvc: for aarch64 Windows devices.
```

  - Hexagon Arch 

>     - Users need to know which architecture is used in their devices

Reviewer 3: ```suggestion
    - When using HTP bakcend, users need to know which architecture is used in their devices
```

      - Please search "Supported Snapdragon devices" in https://docs.qualcomm.com/doc/80-63442-10/topic/QNN_general_overview.html and find the "Hexagon Arch" column to get the Hexagon Arch of the target device
      - For example, if you want to run on a Snapdragon 8 Elite Gen 5 (SM8850), the Hexagon Arch is V81
  - QNN libraries
    - After we know above information, users can locate all required libraries to execute QNN:
      - ${QAIRT}/lib/{Plaform}/libQnnSystem.so

>       - ${QAIRT}/lib/{Plaform}/libQnn{Backend}*.so

Reviewer 2: ```suggestion
      - ${QAIRT}/lib/{Platform}/libQnnSystem.so
      - ${QAIRT}/lib/{Platform}/libQnn{Backend}*.so
```

      - ${QAIRT}/lib/hexagon-v{HexagonArch}/unsigned/libQnnHtpV{HexagonArch}Skel.so
    - For example, users want to execute QNN HTP backends on SM8850, they need to find the following files:
      - ${QAIRT}/lib/aarch64-android/libQnnSystem.so
      - ${QAIRT}/lib/aarch64-android/libQnnHtp.so
      - ${QAIRT}/lib/aarch64-android/libQnnHtpV81Stub.so

>       - ${QAIRT}/lib/hexagon-v81/unsigned/libQnnHtpV81Skel.so

Reviewer 3: I think we can list all required libraries here, not only for execution but also for compilation.
```suggestion
      - ${QAIRT}/lib/hexagon-v81/unsigned/libQnnHtpV81Skel.so
      - ${QAIRT}/lib/hexagon-v81/unsigned/libQnnHtpPrepare.so
```

>     - For exmaple, users want to compiled model for SM8850 with HTP backend in linux work station, they need to find the following files:

Reviewer 2: ```suggestion
    - For example, users want to compiled model for SM8850 with HTP backend in linux work station, they need to find the following files:
```

      - ${QAIRT}/lib/x86_64-linux-clang/libQnnHtp.so

>       - ${QAIRT}/lib/x86_64-linux-clang/libQnnPrepare.so

Reviewer 3: ```suggestion
```

Modify ${LiteRT}/third_party/qairt/workspace.bzl and change `strip_prefix`& `url`. Please get the url of QAIRT SDK from https://softwarecenter.qualcomm.com/catalog/item/Qualcomm_AI_Runtime_Community.

For example, the original workspace.bzl is using "qairt/2.42.0.251225" and users want to use 2.44, then the workspace.bzl should be changed from:
```bash
def qairt():
    configurable_repo(
        name = "qairt",
        build_file = "@//third_party/qairt:qairt.BUILD",
        local_path_env = "LITERT_QAIRT_SDK",
        strip_prefix = "qairt/2.42.0.251225",
        url = "https://softwarecenter.qualcomm.com/api/download/software/sdks/Qualcomm_AI_Runtime_Community/All/2.42.0.251225/v2.42.0.251225.zip",
        file_extension = "zip",
    )
```
To:
```bash
def qairt():
    configurable_repo(
        name = "qairt",
        build_file = "@//third_party/qairt:qairt.BUILD",
        local_path_env = "LITERT_QAIRT_SDK",
        strip_prefix = "qairt/2.44.0.260225",
        url = "https://softwarecenter.qualcomm.com/api/download/software/sdks/Qualcomm_AI_Runtime_Community/All/2.44.0.260225/v2.44.0.260225.zip",
        file_extension = "zip",
    )
```

> ## Bazel Flow (Compile & Execute)

Reviewer 3: ```suggestion
## Example: Compile and Execute a Model (through Bazel)
```

> Build "apply_plugin_main" compilation tool and compile the model on Linux workstation (Ubuntu-22.04):

Reviewer 3: ```suggestion
### AOT compilation on host machine(x86 Linux)
Build "apply_plugin_main" compilation tool and compile the model on host machine (take x86 Ubuntu-22.04 for example):
```

### Steps
```bash
# AOT flow
# Build apply_plugin_main with bazel
bazel build -c opt --cxxopt=--std=c++17 --nocheck_visibility //litert/tools:apply_plugin_main

# Execute apply_plugin_main to compile model
# For example: export LD_LIBRARY_PATH=<PATH_TO_LITERT_REPO>/third_party/qairt/qairt/2.42.0.251225/lib/x86_64-linux-clang
export LD_LIBRARY_PATH=<PATH_TO_UNZIPPED_QAIRT_SDK>/lib/x86_64-linux-clang

# For "--soc_model", the users should set the soc_model based on the target device
# Please search "Supported Snapdragon devices" in https://docs.qualcomm.com/doc/80-63442-10/topic/QNN_general_overview.html to find the correct "Snapdragon Device/Chip" column for your device.

> # For exmaple: bazel-bin/litert/tools/apply_plugin_main --cmd apply --libs bazel-bin/litert/vendors/qualcomm/compiler --soc_model SM8850 --soc_manufacturer Qualcomm --model /data/local/tmp/source_model.tflite -o /data/local/tmp/compile_model.tflite

Reviewer 2: ```suggestion
# For example: bazel-bin/litert/tools/apply_plugin_main --cmd apply --libs bazel-bin/litert/vendors/qualcomm/compiler --soc_model SM8850 --soc_manufacturer Qualcomm --model /data/local/tmp/source_model.tflite -o /data/local/tmp/compile_model.tflite
```

> bazel-bin/litert/tools/apply_plugin_main --cmd apply --libs bazel-bin/litert/vendors/qualcomm/compiler --soc_model <SOC_MODEL_NAME> --soc_manufacturer Qualcomm --model <PATH_TO_YOUR_.TFLITE_MODEL> -o <PATH_TO_YOUR_COMPILED_.TFLITE_MODEL>

Reviewer 3: ```suggestion
${LiteRT}/bazel-bin/litert/tools/apply_plugin_main --cmd apply --libs ${LiteRT}/bazel-bin/litert/vendors/qualcomm/compiler --soc_model <SOC_MODEL_NAME> --soc_manufacturer Qualcomm --model <PATH_TO_YOUR_.TFLITE_MODEL> -o <PATH_TO_YOUR_COMPILED_.TFLITE_MODEL>
```

Reviewer 3: add `cd` or setup environment variable, or assume use has already setup required paths.

> ```

Reviewer 3: Prefer move most comments in shell script outside codeblock, and make codeblock become a clear example.

> Build "run_model" and required libraries on Linux workstation:

Reviewer 3: ```suggestion
### Execution on target machine(Android)
Build "run_model" and required libraries on host machine(x86 Linux):
```

```bash
# Build run_model, dispatch_api_so and qnn_compiler_plugin_so for Android device
bazel build -c opt --cxxopt=--std=c++17 --nocheck_visibility --config=android_arm64 --copt=-DABSL_FLAGS_STRIP_NAMES=0 //litert/tools:run_model
bazel build -c opt --cxxopt=--std=c++17 --nocheck_visibility --config=android_arm64 --copt=-DABSL_FLAGS_STRIP_NAMES=0 //litert/vendors/qualcomm/dispatch:dispatch_api_so
```

Execute the compiled model on Android device with QNN HTP backend:
```bash
# Upload required QNN libraries from Linux workstation to Android device
$ <PATH_TO_UNZIPPED_QAIRT_SDK>/lib/aarch64-android/libQnnSystem.so
$ <PATH_TO_UNZIPPED_QAIRT_SDK>/lib/aarch64-android/libQnnHtp.so
$ <PATH_TO_UNZIPPED_QAIRT_SDK>/lib/aarch64-android/libQnnHtpV81Stub.so
$ <PATH_TO_UNZIPPED_QAIRT_SDK>/lib/hexagon-v81/unsigned/libQnnHtpV81Skel.so

# Upload required LiteRT files from Linux workstation to Android device
$ <PATH_TO_LITERT_REPO>/bazel-bin/litert/c/libLiteRt.so
$ <PATH_TO_LITERT_REPO>/bazel-bin/litert/vendors/qualcomm/dispatch/libLiteRtDispatch_Qualcomm.so
$ <PATH_TO_LITERT_REPO>/bazel-bin/litert/tools/run_model

> # Upload the model compiled by previouse step from Linux workstation to Android device

Reviewer 2: ```suggestion
# Upload the model compiled by previous step from Linux workstation to Android device
```

$ <PATH_TO_YOUR_COMPILED_.TFLITE_MODEL> 

# Set environment variables on Android device
export LD_LIBRARY_PATH=<PATH_TO_TEST_FOLDER>
export ADSP_LIBRARY_PATH=<PATH_TO_TEST_FOLDER>

# Execute the model on Android device, please use "./run_model --helpfull" to find all Qualcomm options and descriptions
# Below is an example of running a model with 50 iterations and QNN HTP performance mode set to "burst"
# The log level is set to "off" to reduce logging overhead

> ./run_model --graph=<PATH_TO_YOUR_COMPILED_.TFLITE_MODEL> --dispatch_library_dir=<PATH_TO_TEST_FOLDER> --iterations 50 --qualcomm_htp_performance_mode burst --qualcomm_log_level off

Reviewer 3: Prefer moving most comments outside codeblock, and make codeblock become a clear example(or script).

```

> ### 

Reviewer 3: ```suggestion
```

> ## CMake Flow (Compile & Execute)

Reviewer 3: ```suggestion
## Example: Compile and Execute a Model (through CMake)
```

There's an official CMake build tutorial in <PATH_TO_LITERT_REPO>/g3doc/instructions/CMAKE_BUILD_INSTRUCTIONS.md for your reference.

> ### Prerequiste

Reviewer 2: ```suggestion
### Prerequisite
```

1. CMake
   - Currently we use 4.0.1.
```bash
wget https://github.com/Kitware/CMake/releases/download/v4.0.1/cmake-4.0.1-linux-x86_64.tar.gz
tar -xzvf cmake-4.0.1-linux-x86_64.tar.gz
```

2. Android NDK for Android build
   - Currently we use android-ndk-r25c.
```bash
wget https://dl.google.com/android/repository/android-ndk-r25c-linux.zip
unzip android-ndk-r25c-linux.zip

> ```

Reviewer 3: I think this is included `Enviroment Setup`, and we can remove them here.

> ### Cmake build x86

Reviewer 3: ```suggestion
### AOT compilation on the host machine(x86 Linux)
```

```bash
cd <PATH_TO_LITERT_REPO>/litert

# Configure, it takes a while to download tensorflow
cmake --preset default

# Build required targets, the built files are located in:
#   ./c/libLiteRt.so
#   ./vendors/qualcomm/compiler/libLiteRtCompilerPlugin_Qualcomm.so
#   ./tools/apply_plugin_main
cmake --build cmake_build --target apply_plugin_main qnn_compiler_plugin litert_runtime_c_api_shared_lib -j8

# Target "extract_bytecode" is a tool used to extract the context binary file from LiteRT compiled .tflite model
# The extracted context binary file can be used to run the model with QNN runtime (qnn-net-run)
# If the users intend to run the model with LiteRT runtime, users could ignore to build this target

> cmake --build cmake_build --target extract_bytecode -j8

Reviewer 3: Suggest move to an independent section about how to extract QNN context binary from a compile .tflite model.

# Set environment variables for QNN libraries

> export LD_LIBRARY_PATH=<PATH_TO_LITERT_REPO>/litert/cmake_build/_deps/qnn_headers/qairt/2.44.0.260225/lib/x86_64-linux-clang

Reviewer 3: ```suggestion
export LD_LIBRARY_PATH=${LiteRT}/litert/cmake_build/_deps/qnn_headers/qairt/2.44.0.260225/lib/x86_64-linux-clang
```

# Execute apply_plugin_main to compile model
cd cmake_build
# For "--soc_model", the users should set the soc_model based on the target device
# Please search "Supported Snapdragon devices" in https://docs.qualcomm.com/doc/80-63442-10/topic/QNN_general_overview.html to find the correct "Snapdragon Device/Chip" column for your device.

> ./tools/apply_plugin_main --cmd apply --libs ./vendors/qualcomm/compiler/ --soc_model <SOC_MODEL_NAME> --soc_manufacturer Qualcomm --model <PATH_TO_YOUR_.TFLITE_MODEL> -o <PATH_TO_YOUR_COMPILED_.TFLITE_MODEL>

Reviewer 3: ```suggestion
${LiteRT}/litert/cmake_build/tools/apply_plugin_main --cmd apply --libs ${LiteRT}/litert/cmake_build/vendors/qualcomm/compiler/ --soc_model <SOC_MODEL_NAME> --soc_manufacturer Qualcomm --model <PATH_TO_YOUR_.TFLITE_MODEL> -o <PATH_TO_YOUR_COMPILED_.TFLITE_MODEL>
```

```

> ### Cmake build Android

Reviewer 3: ```suggestion
### Execution on target machine(Android)
```

Below is based on android-ndk-r25c-linux.

```bash
export ANDROID_NDK_HOME=<PATH_TO_ANDROID_NDK>

cd <PATH_TO_LITERT_REPO>/litert

# Configure, it takes a while to build flatc and download tensorflow
cmake --preset android-arm64

# Build required targets, the built files are located in:
#   ./c/libLiteRt.so
#   ./vendors/qualcomm/dispatch/libLiteRtDispatch_Qualcomm.so
#   ./tools/run_model
cmake --build cmake_build_android_arm64 --target run_model litert_runtime_c_api_shared_lib dispatch_api_qualcomm_so -j8

# Upload required QNN libraries from Linux workstation to Android device
$ <PATH_TO_UNZIPPED_QAIRT_SDK>/lib/aarch64-android/libQnnSystem.so
$ <PATH_TO_UNZIPPED_QAIRT_SDK>/lib/aarch64-android/libQnnHtp.so
$ <PATH_TO_UNZIPPED_QAIRT_SDK>/lib/aarch64-android/libQnnHtpV81Stub.so
$ <PATH_TO_UNZIPPED_QAIRT_SDK>/lib/hexagon-v81/unsigned/libQnnHtpV81Skel.so

# Upload required LiteRT files from Linux workstation to Android device
$ <PATH_TO_LITERT_REPO>/litert/cmake_build_android_arm64/c/libLiteRt.so
$ <PATH_TO_LITERT_REPO>/litert/cmake_build_android_arm64/vendors/qualcomm/dispatch/libLiteRtDispatch_Qualcomm.so
$ <PATH_TO_LITERT_REPO>/litert/cmake_build_android_arm64/tools/run_model

> # Upload the model compiled by previouse step from Linux workstation to Android device

Reviewer 2: ```suggestion
# Upload the model compiled by previous step from Linux workstation to Android device
```

$ <PATH_TO_YOUR_COMPILED_.TFLITE_MODEL> 

# Set environment variables on Android device
export LD_LIBRARY_PATH=<PATH_TO_TEST_FOLDER>
export ADSP_LIBRARY_PATH=<PATH_TO_TEST_FOLDER>

# Execute the model on Android device, please use "./run_model --helpfull" to find all Qualcomm options and descriptions
# Below is an example of running a model with 50 iterations and QNN HTP performance mode set to "burst"
# The log level is set to "off" to reduce logging overhead

> ./run_model --graph=<PATH_TO_YOUR_COMPILED_.TFLITE_MODEL> --dispatch_library_dir=<PATH_TO_TEST_FOLDER> --iterations 50 --qualcomm_htp_performance_mode burst --qualcomm_log_level off

Reviewer 3: Make them become complete script, e.g. use adb.

```

### Cmake build cross compilation oe-linux for IoT device

> The cross compilation is only required when the host machine is x86 and the target machine is aarch64 based.

Reviewer 3: ```suggestion
The cross compilation is only required when the host machine is x86 Linux and the target machine is aarch64 Linux based.
```

Below is a sample with x86 linux host machine and build for IQ-8275 (Aarch64) device.
- This step needs to download the eSDK for your target IoT device.

```bash

> # Prerequiste for Linux workstation

Reviewer 2: ```suggestion
# Prerequisite for Linux workstation
```

sudo locale-gen en_US.UTF-8
sudo apt update
sudo apt install gawk
# Configure Android to generate host_flatc_build
cd <PATH_TO_LITERT_REPO>/litert

# Check the META version on the IQ-8275
uname -a
# Find the corresponding eSDK version in https://artifacts.codelinaro.org/ui/native/qli-ci/flashable-binaries/qimpsdk/, select "qcs8275-iq-8275-evk-pro-sku/"

> #   If the host machine is arm-based, please use the link with the prefix “arm”

Reviewer 3: ```suggestion
```

export LINUX_AARCH64_ESDK=<PATH_TO_INSTALLED_ESDK>
cd <PATH_TO_LITERT_REPO>/litert

# Configure, it takes a while to build flatc and download tensorflow
cmake --preset linux-aarch64-iq8

# Build required targets, the built files are located in:
#   ./c/libLiteRt.so
#   ./vendors/qualcomm/dispatch/libLiteRtDispatch_Qualcomm.so
#   ./tools/run_model
cmake --build cmake_build_linux_aarch64_8275/ --target run_model litert_runtime_c_api_shared_lib dispatch_api_qualcomm_so -j8

# Upload required QNN libraries from Linux workstation to IoT device
$ <PATH_TO_UNZIPPED_QAIRT_SDK>/lib/aarch64-oe-linux-gcc11.2/libQnnSystem.so
$ <PATH_TO_UNZIPPED_QAIRT_SDK>/lib/aarch64-oe-linux-gcc11.2/libQnnHtp.so
$ <PATH_TO_UNZIPPED_QAIRT_SDK>/lib/aarch64-oe-linux-gcc11.2/libQnnHtpV75Stub.so
$ <PATH_TO_UNZIPPED_QAIRT_SDK>/lib/hexagon-v75/unsigned/libQnnHtpV75Skel.so

# Upload required LiteRT files from Linux workstation to IoT device
$ <PATH_TO_LITERT_REPO>/litert/cmake_build_linux_aarch64_8275/c/libLiteRt.so
$ <PATH_TO_LITERT_REPO>/litert/cmake_build_linux_aarch64_8275/vendors/qualcomm/dispatch/libLiteRtDispatch_Qualcomm.so
$ <PATH_TO_LITERT_REPO>/litert/cmake_build_linux_aarch64_8275/tools/run_model

> # Upload the model compiled by previouse step from Linux workstation to IoT device

Reviewer 2: ```suggestion
# Upload the model compiled by previous step from Linux workstation to IoT device
```

> # Need to use --soc_model QCS8300 for model compialtion for IQ-8275

Reviewer 2: ```suggestion
# Need to use --soc_model QCS8300 for model compilation for IQ-8275
```

$ <PATH_TO_YOUR_COMPILED_.TFLITE_MODEL> 

# Execute the model on Android device, please use "./run_model --helpfull" to find all Qualcomm options and descriptions
# Below is an example of running a model with 50 iterations and QNN HTP performance mode set to "burst"
# The log level is set to "off" to reduce logging overhead
# Some error logs for IQ-8275 execution are expected
./run_model --graph=<PATH_TO_YOUR_COMPILED_.TFLITE_MODEL> --dispatch_library_dir=<PATH_TO_TEST_FOLDER> --iterations 50 --qualcomm_htp_performance_mode burst --qualcomm_log_level off
```

## Bringup IoT device

### Flash oe-linux
The official document of IQ-8275 (oe-linux) is here: https://docs.qualcomm.com/doc/80-80020-281/topic/iq8-ug-update-the-sw.html for your reference.

> ```bash

Reviewer 3: Suggest split this code block into multiple blocks.

wget https://apigwx-aws.qualcomm.com/qsc/public/v1/api/download/software/tools/Qualcomm_Device_Loader/Linux/Debian/2.5.0/QDL_2.5_Linux_x64.zip
unzip QDL_2.5_Linux_x64.zip

# Find "USB0 Type-C port" in https://docs.qualcomm.com/doc/80-80020-281/topic/iq8-ug-introduction.html
# Connect the device through "USB0 Type-C port" to the host machine
# UFS provisioning and reboot device
wget https://artifacts.codelinaro.org/artifactory/codelinaro-le/Qualcomm_Linux/QCS8300/provision.zip
unzip provision.zip
<QDL_TOOL_PATH>/qdl --storage ufs prog_firehose_ddr.elf provision_1_3.xml
# CDT flashing and reboot device
wget https://artifacts.codelinaro.org/artifactory/codelinaro-le/Qualcomm_Linux/QCS8300/cdt/qcs8275-iq-8275-evk-pro-sku.zip
unzip qcs8275-iq-8275-evk-pro-sku.zip
<QDL_TOOL_PATH>/qdl prog_firehose_ddr.elf rawprogram3.xml patch3.xml

# Flash Safety Island (SAIL)
cd <UNZIPPED_IOT_IMAGE_PATH>/target/qcs8275-iq-8275-evk-pro-sku/qcom-multimedia-image/sail_nor
<QDL_TOOL_PATH>/qdl --storage spinor prog_firehose_ddr.elf rawprogram0.xml patch0.xml
# Flash image

> cd <UNZIPPED_IOT_IMAGE_PATH>//target/qcs8275-iq-8275-evk-pro-sku/qcom-multimedia-image

Reviewer 2: ```suggestion
cd <UNZIPPED_IOT_IMAGE_PATH>/target/qcs8275-iq-8275-evk-pro-sku/qcom-multimedia-image
```
